Refactoring io preprocessing

GWeindel opened the pull request:
Hi @jelmerborst , @lvmaanen @rickdott @qubixes @maartenschermer 
I made a quite big PR (oups). I guess none of you has the time to read it all so I just tag each of you on the parts that will probably interest you/needs your review.

# Major changes
1. The EEG/MEG data is now stored for the whole epochs instead of being cut at RT when the data is read using `read_mne_data`. That allows to more easily plot ERPs but also switch between different durations without re-reading all the data  @jelmerborst, @rickdott
2. The `preprocessing` has been renamed to `transformers` to make the distinction with EEG/MEG preprocessing, typically done before HMP. The new classes follow the modular organization as in `models` @qubixes @maartenschermer
3. The filtering option was removed from the transformers as there are a few users who were using it wrongly at this stage and the maintenance is tricky @jelmerborst 
4. Cutting the data at the RT (or more generally `interval` now) is performed as part of the new `transformers` thus the arguments that applied to this operation were moved and went from `read_mne_data` to the `transformers` classes. 
5. Zscoring is now always performed at the trial level without any other option (amongst others to avoid large RAM usage) @lvmaanen @jelmerborst This is because we expect the same sequence in all trials, thus that's the only reasonable transformation given this assumption. Note that this does not mean that we remove the difference, e.g. the channel values at the events' time (typically computed using the original epoched data) won't be affected by this zscoring strategy. Furthermore Issue #57 shows that this is the best strategy for simulated data at least (and real data from my experience)
6. The PCA is now computed without sklearn to avoid the dependency and is applied on the average covariance matrix with a small regularization. These changes respectively gives more control and slightly improves performance
7. MCCA was removed for now as it is too experimental but might be re-implemented later on @jelmerborst 
8. The result of preprocessed data is a stacked representation with trials X components X samples instead of all_samples X components which wasn't easy to use for others.


# Overall consequences:
- The interval cut (as well as providing metadata) is now optional and the data read is the actual data stored as epochs
- The distinction between preprocessed data and trial data is much clearer. 
- The different projections method now share all the same structure
- The transforming step now uses 2x less RAM and is also slightly faster (mostly due to the change of `float64` to `float32` and the removal of large array operations in `transformers`)

# Coming 
In another PR I will further refactor the `io` module to allow for more flexibility in the data type being read and also to avoid very long functions with unclear paths

GWeindel commented:
The tests are broken, I'll fix when we agree on the new revision and add some tests to the new structure

qubixes commented on hmp/transformers/common.py:
```diff
@@ -0,0 +1,79 @@
+"""Transforms epoched data using a a custom linear combination for HMP analysis.
```
```suggestion
"""Transforms epoched data using a custom linear combination for HMP analysis.
```

qubixes commented on hmp/transformers/custom.py:
Change filename to lowercase (drop the proj?).

qubixes commented on hmp/transformers/common.py:
```diff
@@ -0,0 +1,79 @@
+        Name of the variable that contains the per-trial intervals in the epoch_data used for cropping.
+    offset_after : float
+        Time offset after interval start for cropping.
+    too_short : float, optional
```
You might want to consider the naming of this variable. I would think it is a boolean. Perhaps min_duration or something like it?

qubixes commented on hmp/transformers/common.py:
```diff
@@ -0,0 +1,79 @@
+        Time offset after interval start for cropping.
+    too_short : float, optional
+        Minimum duration threshold for keeping epochs.
+    too_long : float, optional
```
Same here.

qubixes commented on hmp/transformers/ProjCustom.py:
```diff
@@ -0,0 +1,79 @@
+        Threshold for rejecting noisy epochs.
+    apply_standard : bool
+        Whether to standardize variance across participants.
+    apply_zscore :bool
```
```suggestion
    apply_zscore : bool
```

qubixes commented on hmp/transformers/identity.py:
Naming

qubixes commented on hmp/transformers/ProjIdentity.py:
```diff
@@ -0,0 +1,86 @@
+        Threshold for rejecting noisy epochs.
+    apply_standard : bool
+        Whether to standardize variance across participants.
+    apply_zscore :bool
```
```suggestion
    apply_zscore : bool
```

qubixes commented on hmp/transformers/identity.py:
```diff
@@ -0,0 +1,86 @@
+            copy=copy,
+        )
+        warn('Identity projection might pose problems of dimensionality'
+             'and collinearity of channels. Thus rendering HMP estimation'
```
```suggestion
             ' and collinearity of channels. Thus rendering HMP estimation'
```

qubixes commented on hmp/transformers/identity.py:
```diff
@@ -0,0 +1,86 @@
+        )
+        warn('Identity projection might pose problems of dimensionality'
+             'and collinearity of channels. Thus rendering HMP estimation'
+             'difficult, use with care!')
```
```suggestion
             ' difficult, use with care!')
```

qubixes commented on hmp/transformers/pca.py:
```diff
@@ -0,0 +1,137 @@
+        pca_weights = xr.DataArray(evecs, dims=("channel", "component"), coords=coords)
+        return pca_weights, eigvals[ix]
+
+    def user_input_n_comp(self, data, n_comp, channels):
```
Data types?

qubixes commented on hmp/transformers/base.py:
```diff
@@ -0,0 +1,198 @@
+        """
+        sfreq=self.sfreq
+        too_short = self.too_short
+        too_long = self.too_long        
```
Is this really necessary? Might be clearer if you just use self.too_short.

qubixes commented on hmp/transformers/base.py:
```diff
@@ -0,0 +1,198 @@
+            too_long = (int(epoch_data.sample.max()) - offset_after_samples)/sfreq
+        if too_short is None:
+            too_short = 1 / sfreq
+        if too_long < 0 or too_short < 0 or too_long < too_short:
```
This should be validated during `__init__`.

qubixes commented on hmp/transformers/base.py:
```diff
@@ -0,0 +1,199 @@
+            else:
+                epoch_data.values[i, :, :] = np.nan
+    
+        assert rej < len(epoch_data), 'All trials rejected, inspect intervals and rejection criterion'
```
Don't use asserts for non-testing code. Also, shouldn't it be the other way around?

GWeindel replied:
It was correct but probably the assert syntax made this unclear

qubixes reviewed (changes requested):
I have a few small comments, looks quite nice overall!

GWeindel commented:
I updated the code based on @qubixes feedback and made passing tests so it's production ready. Note that I also removed zscoring completely and only "whiten" the components by dividing them with their std computed on all samples. @rickdott @jelmerborst @lvmaanen  any additional comments or suggestions? I'm thinking to merge on monday

rickdott commented:
I'll look into the part you tagged me in tomorrow, quick check: did you also update the tutorials/docs to use the new preprocessing/transformers API? Dont see it in the changed files

GWeindel commented:
> I'll look into the part you tagged me in tomorrow, quick check: did you also update the tutorials/docs to use the new preprocessing/transformers API? Dont see it in the changed files

I updated the tutorials, to look at the API you need to build it
```bash
pip install -e ".[docs]"
cd docs; make clean html
```
And then open the index.html in docs/build/html.

rickdott commented on hmp/transformers/pca.py:
```diff
@@ -0,0 +1,156 @@
+        interval_id: str = 'rt',
+        offset_after_end: float = 0,
+        min_duration: float = 0,
+        max_duration: float =float('Inf'),
```
Forgot a space here

rickdott replied:
max_duration: float = float('Inf'),

rickdott commented on hmp/transformers/base.py:
```diff
@@ -0,0 +1,208 @@
+import xarray as xr
+
+
+class BaseTransformer(ABC):
```
I dont 100% see the use of this being a class rather than separate functions, maybe it could be changed to take epoch_data in its initialization, we could then run self.common_preprocess() on that within the BaseTransformer.__init__(), since we run that function in every child of BaseTransformer

rickdott commented on hmp/transformers/base.py:
```diff
@@ -0,0 +1,208 @@
+
+    Methods
+    -------
+    common_preprocess()
```
common_preprocess also takes a data argument (currently, see prev comment)

rickdott commented on hmp/trialdata.py:
```diff
@@ -43,35 +44,31 @@ class TrialData:
         ----------
-        preprocessed : Preprocessing or xr.DataArray
-            The preprocessed object or xarray DataArray containing the preprocessed data.
+        transformed : BaseTransfromer or xr.DataArray
```
typo (BaseTransfromer)

GWeindel commented:
@rickdott I finished, documentation is also updated, I only need to rework on the tests but you can try if you have time, default should be fine

GWeindel commented:
The PR is ready to be integrated

Of note is that the default behavior changed slightly compared to what I wrote in the beginning, by default data is: 	
	- centered (mainly for covariance computation) 
	- the trials are standardized (across components) -> All trials are contributing equally
	- the participants are zscored (over components) -> All components are re-scaled for each participant as they might not all have the same SNR
	- SVD is explicitly used rather than PCA (but was already the case before this PR as it was the option passed to sklearn)

rickdott commented:
Looks good, I dont think there are review-worthy changes in the commits since I last took a look.
